### PR TITLE
Capture modularisation/cleanup

### PR DESCRIPTION
**Pull request by johnnovak**
Hacking'n'slashing the entangled spaghetti mess responsible for capturing stuff into more manageable pieces.
No functional changes, just reorganising stuff & cleanup. AFAIK I haven't broken anything, only upset the linters a bit... (to be fixed).

This all lived in `hardware.cpp` previously in a big cosy ball of mud 😩 Sure, that's a good place... we're all dealing with "hardware" after all... computers are hardware too, and so on 🤦🏻 🥶 

This is in preparation for the screenshot improvements.

### CODE REVIEW

**Comment by kcgen** [comment]
> This all lived in hardware.cpp previously in a big cosy ball of mud weary Sure, that's a good place... we're all dealing with "hardware" after all... computers are hardware too, and so on 🤦🏻 :cold_face: 

:rofl:

**Comment by ThomasEricB** [comment]
Nice work @johnnovak ! You really showed `hardware.cpp` who is boss!

**Comment by johnnovak** [comment]
> You really showed `hardware.cpp` who is boss!

It had it coming. 😎

**Comment by johnnovak** [comment]
So, for the record, this builds fine on macOS, even after a full clean & rebuild. I'll screw around on Windows today or tomorrow to eradicate those build failures—hopefully that will get rid of most of the Linux failures too.

File: include/hardware.h [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- *  Copyright (C) 2002-2021  The DOSBox Team
+ *  Copyright (C) 2002-2023  The DOSBox Team
```
Comment (dreamer) [inline include/hardware.h:2]: We are keeping "DOSBox Team" copyrights intact; in cases like this one, add another copyright line for "DOSBox Staging Team" (grep through the code will give plenty of examples).

File: src/capture/capture.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+		*test = 0;
+		if (strncasecmp(tempname, file_start, strlen(file_start)) != 0)
+			continue;
+		const int num = atoi(&tempname[strlen(file_start)]);
+		if (num >= last)
+			last = num + 1;
+	}
+	close_directory(dir);
+	char file_name[CROSS_LEN];
+	sprintf(file_name,
```
Comment (dreamer) [inline src/capture/capture.cpp:91]: Let's update it to `safe_sprintf` while we're at it.

File: include/hardware.h (hunk 1 above)
Comment (johnnovak) [reply include/hardware.h:2]: Yeah okay. TBH I did not even read it, just updated the date 😄

File: src/capture/capture.cpp (hunk 2 above)
Comment (johnnovak) [reply src/capture/capture.cpp:91]: Sure, good idea.

File: src/capture/capture.cpp [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+	return file_name;
+}
+
+FILE* CAPTURE_CreateFile(const char* type, const char* ext)
+{
+	const auto file_name = capture_generate_filename(type, ext);
+
+	FILE* handle = fopen(file_name.c_str(), "wb");
+	if (handle) {
+		LOG_MSG("Capturing %s to %s", type, file_name.c_str());
```
Comment (dreamer) [inline src/capture/capture.cpp:107]: Let's surround paths we print to log with quotes: `'%s'`.

Comment (johnnovak) [reply src/capture/capture.cpp:107]: Good idea.

**Comment by johnnovak** [comment]
Moving this back to draft as I'm going to clean it up more substantially in addition to the fairly mechanical splitting/refactoring/etc. I've already done.

The end goal is to make all this easily modifiable for future enhancements... because currently it isn't 😩

**Comment by johnnovak** [comment]
Okay, this is ready for final review @kcgen & @dreamer. I've regressions tested everything and all capture functionality works 100% fine.

I think the whole thing should be condensed into a single commit; the commit history doesn't add much value as files have been split, moved around, and refactored in a major way. Plus I did break a few things in the first phase because the state handling was still a mess... This has been rectified in my last big refactor commit. Now you can mostly follow what's going on by reading the code, finally 😅

File: include/hardware.h (hunk 1 above)
Comment (johnnovak) [reply include/hardware.h]: Done in all files.

File: src/capture/capture.cpp (hunk 3 above)
Comment (johnnovak) [reply src/capture/capture.cpp]: Done.

File: src/capture/capture.cpp (hunk 2 above)
Comment (johnnovak) [reply src/capture/capture.cpp]: Done.

**Comment by johnnovak** [comment]
Someone give me a tick please? @kcgen @dreamer @FeralChild64 Feel free to test, I could not spot any regressions/problems after the refactor.

**Comment by kcgen** [comment]
The "Move parts of generate rendered screenshot to capture_image" commit fails to compile (bad for bisecting):

``
In file included from ../../include/sdlmain.h:33,
                 from ../../src/capture/capture_image.cpp:24:
../../src/capture/../libs/ppscale/ppscale.h: In constructor ‘PPScale::PPScale(int, int, double, bool, int, int)’:
../../src/capture/../libs/ppscale/ppscale.h:51:17: error: ‘assert’ was not declared in this scope
   51 |                 assert(source_w > 0);
      |                 ^~~~~~
```

**Review by kcgen — COMMENTED** [review]
Tests are running well - I tested wav, avi, and both screenshots all individually and then all combinations:
wav
avi
screenshots
wav + avi
wav + screenshots
wav + avi + screenshots
avi + screenshots

So if this is cleanup only without functional changes intended, then I think it's OK.  Just need all commits building and it'll be set!

**Comment by johnnovak** [comment]
Thanks for testing it @kcgen. I've squashed the offending commits, I'm pretty sure I was able to compile every commit on macOS, but they I was relying on incremental commits.

**Comment by kcgen** [comment]
> Thanks for testing it @kcgen. I've squashed the offending commits, I'm pretty sure I was able to compile every commit on macOS, but then I was relying on incremental compilation.

:+1: , your good commit messages indicate unbuildable commits still exist in the latest push:



Checking prior-to-fixed macOS commit -- indeed, it doesn't build:

**Review by kcgen — APPROVED** [review]
approving - feel free to merge @johnnovak once those last compile commits are handled :rocket:

**Comment by kcgen** [comment]
> I think the whole thing should be condensed into a single commit

Sorry I missed that earlier! Given it all checks out, I'll go-ahead and squash and merge.

**Comment by johnnovak** [comment]
> > I think the whole thing should be condensed into a single commit
> 
> Sorry I missed that earlier! Given it all checks out, I'll go-ahead and squash and merge.

Yeah, by the final commit everything ended up being substantially refactored anyway, so hence my suggestion to just squash the whole thing. It would be possible to do such big refactors and keep all builds green on the different plstforms, but that's much more time consuming... Thanks for merging.

**Pull request merged**